Shopify/quilt
# Remove loom and replace it with calling rollup directly

Follow up to #2550, #2555 #2556.

This PR removes the final bit of loom we were depending on. It replaces using loom for builds (`yarn run build`) with using rollup directly. We define `rollup.config.mjs` files in each package, that all use a config builder function that is defined in `config/rollup.mjs`.

The change looks big but almost all of it is "delete 72 packages/*/loom.config.ts,  and replace them with 72 new `packages/*/rollup.config.mjs` files.

The rollup plugin that generates entrypoints is kinda gnarly but it works so whatever. It turns out about 75% of the packages in the repo only have a single entrypoint so entrypoint generation for them is overkill - as `main` in their package.json could reference the expected place, but I don't to be dealing with that now. The Aim of this PR is to produce identical output and that's what it's doing (with 4 slight exceptions where we've fixed a false-positive bug in loom).

To generate rollup builds for all files you can run `yarn workspaces run rollup -c`, however this takes about 60 seconds as it sequentially starts up a new node process for each package. I have created a `scripts/build.mjs` file that runs the rollup build for each package in parallel within the same process. This reduction in overhead means a full build takes about 10 seconds (the same as loom, which did something similar).

I have removed all remaining mentions of loom, including the `loom.config.ts` files at the root and in each package and have renamed the root level `.loom` folder to `.cache`.


### To test

Compare build output to previous builds:

- Check out the main branch of quilt in your `~/projects/quilt` directory and run `yarn`, `yarn clean && yarn build` to produce a clean build
- Check out this branch of quilt in your `~/src/github.com/Shopify/quilt` directory and run `yarn`, `yarn clean && yarn build` to produce a clean build
- Run `for PACKAGENAME in $(ls -1 packages); diff -ru ~/projects/quilt/packages/$PACKAGENAME packages/$PACKAGENAME -x '*.tsbuildinfo' -x '*.d.ts.map'  -x 'rollup.config.mjs' -x 'loom.config.ts'` to compare the build folder output.
  -  Note that there are 4 differences. The entrypoint files `packages/react-i18n/generate-dictionaries.esnext`, `packages/react-i18n/generate-dictionaries.mjs`, `packages/react-i18n/generate-index.esnext` and `packages/react-i18n/generate-index.mjs` all lack a `export {default}` line. This new output is correct as there is no default export from the files it references. This was erroniously added in loom because those source files contained the string `export default` in a place that wasn't exporting content.
